firewood/src/merkle/node.rs

```diff
@@ -717,6 +718,176 @@ impl Storable for Node {
+                    if let Some(c) = c {
+                        if c.len() >= TRIE_HASH_LEN {
+                            let hash = TrieHash(Keccak256::digest(c).into());
+                            let serialized_hash = other
+                                .serialize(&&hash[..])
+                                .map_err(|e| S::Error::custom(format!("bincode error: {e}")))?;
+                            s[i] = Encoded::Data(serialized_hash);
+                        } else {
+                            s[i] = Encoded::Raw(c.to_vec());
+                        }
+                    }
+                }
+                if let Some(Data(val)) = &value {
+                    let serialized_val = other
+                        .serialize::<Vec<u8>>(val)
+                        .map_err(|e| S::Error::custom(format!("bincode error: {e}")))?;
+                    s[MAX_CHILDREN] = Encoded::Data(serialized_val);
+                }
+
+                let mut seq = serializer.serialize_seq(Some(s.len()))?;
+                for e in s {
+                    seq.serialize_element(&e)?;
+                }
+                seq.end()
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for EncodedNode<bincode::DefaultOptions> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let other = bincode::DefaultOptions::new();
+        let items: Vec<Encoded<Vec<u8>>> = Deserialize::deserialize(deserializer)?;
+        let len = items.len();
+        match len {
+            LEAF_NODE_SIZE => {
+                let mut items = items.into_iter();
+                let path = match items.next().unwrap() {
+                    Encoded::Raw(path) => path,
+                    Encoded::Data(path) => Err(D::Error::custom(format!(
+                        "incorrect encoded type for leaf node path {:?}",
+                        path
+                    )))?,
+                };
+                let data = match items.next().unwrap() {
+                    Encoded::Raw(data) => data,
+                    Encoded::Data(data) => Err(D::Error::custom(format!(
+                        "incorrect encoded type for leaf node data {:?}",
+                        data
+                    )))?,
+                };
+                let path_nibbles = Nibbles::<0>::new(&path);
+                let path = PartialPath::from_nibbles(path_nibbles.into_iter()).0;
+                let node = EncodedNodeType::Leaf(LeafNode {
+                    path,
+                    data: Data(data),
+                });
+                Ok(Self {
+                    node,
+                    phantom: PhantomData,
+                })
+            }
+            BRANCH_NODE_SIZE => {
+                let mut items = items;
+
+                // we've already validated the size, that's why we can safely unwrap
+                let data = match items.pop().unwrap() {
```

Comment:
I think I may have written code like this before, but it would be better to just use an iterator and get the children, then the data.